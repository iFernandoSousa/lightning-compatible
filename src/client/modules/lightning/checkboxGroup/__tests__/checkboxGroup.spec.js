/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { createElement } from 'lwc';
import Element from 'c/checkboxGroup';
import {
    verifyClassSet,
    shadowQuerySelector,
    shadowQuerySelectorAll
} from 'lightning/testUtils';

const options = [
    { label: 'Ross', value: 'option1' },
    { label: 'Rachel', value: 'option2' }
];

const defaultProps = {
    name: 'testGroup',
    label: 'testGroup',
    options,
    value: []
};

const createCheckboxGroup = props => {
    const element = createElement('c-checkbox-group', { is: Element });
    Object.assign(element, defaultProps, props);
    document.body.appendChild(element);
    return element;
};

describe('c-checkbox-group', () => {
    afterEach(() => {
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    it('default', () => {
        const element = createCheckboxGroup();
        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    it('one selected option', () => {
        const element = createCheckboxGroup({
            value: ['option1']
        });

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    it(`should have 'fieldset' element with 'aria-required' attribute set to 'false' by default`, () => {
        const element = createCheckboxGroup({
            value: []
        });

        return Promise.resolve().then(() => {
            const fieldsetElement = shadowQuerySelector(element, 'fieldset');

            expect(fieldsetElement.getAttribute('aria-required')).toBe('false');
        });
    });

    it(`should have 'fieldset' element with 'aria-required' attribute set to 'true' when group is required`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: true
        });

        return Promise.resolve().then(() => {
            const fieldsetElement = shadowQuerySelector(element, 'fieldset');

            expect(fieldsetElement.getAttribute('aria-required')).toBe('true');
        });
    });

    it(`should have 'abbr' element with 'title' attribute set to 'required' when group is required`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: true
        });

        return Promise.resolve().then(() => {
            const abbrElement = shadowQuerySelector(element, 'abbr');

            expect(abbrElement).toBeDefined();
            expect(abbrElement.getAttribute('title')).toBe('required');
            expect(abbrElement.textContent).toBe('*');
        });
    });

    it(`should have 'input' elements with no 'required' attributes when group is required`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: true
        });

        return Promise.resolve().then(() => {
            const inputElements = shadowQuerySelectorAll(element, 'input');
            const inputElementsLength = inputElements.length;

            for (let i = 0; i < inputElementsLength; i++) {
                expect(inputElements[i].getAttribute('required')).toBeNull();
            }
        });
    });

    it(`should have 'fieldset' element with 'aria-required' attribute set to 'false' when the group is not required`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: false
        });

        return Promise.resolve().then(() => {
            const fieldsetElement = shadowQuerySelector(element, 'fieldset');

            expect(fieldsetElement.getAttribute('aria-required')).toBe('false');
        });
    });

    it(`should not have 'abbr' element when group is not required`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: false
        });

        return Promise.resolve().then(() => {
            const abbrElement = shadowQuerySelector(element, 'abbr');

            expect(abbrElement).toBeNull();
        });
    });

    it(`should have 'input' elements with no 'required' attributes when group is not required`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: false
        });

        return Promise.resolve().then(() => {
            const inputElements = shadowQuerySelectorAll(element, 'input');
            const inputElementsLength = inputElements.length;

            for (let i = 0; i < inputElementsLength; i++) {
                expect(inputElements[i].getAttribute('required')).toBeNull();
            }
        });
    });

    it(`should have 'fieldset' element with 'aria-required' attribute set to 'false' after the 'required' status is updated`, () => {
        const element = createCheckboxGroup({
            value: [],
            required: true
        });

        element.required = false;

        return Promise.resolve().then(() => {
            const fieldsetElement = shadowQuerySelector(element, 'fieldset');

            expect(fieldsetElement.getAttribute('aria-required')).toBe('false');
        });
    });

    it('disabled', () => {
        const element = createCheckboxGroup({
            value: ['option1'],
            disabled: true
        });

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    it('required and disabled', () => {
        const element = createCheckboxGroup({
            value: ['option1'],
            required: true,
            disabled: true
        });

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    it('value change from one option to the other', () => {
        const element = createCheckboxGroup({
            value: ['option1']
        });

        return Promise.resolve()
            .then(() => {
                const inputs = shadowQuerySelectorAll(element, 'input');
                expect(inputs[0].checked).toBe(true);
                expect(inputs[1].checked).toBe(false);
                element.value = ['option2'];
            })
            .then(() => {
                const inputs = shadowQuerySelectorAll(element, 'input');
                expect(inputs[0].checked).toBe(false);
                expect(inputs[1].checked).toBe(true);
            });
    });

    it('form element class for default variant', () => {
        const element = createCheckboxGroup({
            value: ['option1']
        });

        return Promise.resolve().then(() => {
            verifyClassSet(element, {
                'slds-form-element': true,
                'slds-form-element_stacked': false,
                'slds-form-element_horizontal': false
            });

            element.variant = 'label-stacked';
            return Promise.resolve().then(() => {
                verifyClassSet(element, {
                    'slds-form-element': true,
                    'slds-form-element_stacked': true,
                    'slds-form-element_horizontal': false
                });
            });
        });
    });
    it('form element class for label-stacked variant', () => {
        const element = createCheckboxGroup({
            value: ['option1'],
            variant: 'label-stacked'
        });

        return Promise.resolve().then(() => {
            verifyClassSet(element, {
                'slds-form-element': true,
                'slds-form-element_stacked': true,
                'slds-form-element_horizontal': false
            });
        });
    });
    it('form element class for label-inline variant', () => {
        const element = createCheckboxGroup({
            value: ['option1'],
            variant: 'label-inline'
        });

        return Promise.resolve().then(() => {
            verifyClassSet(element, {
                'slds-form-element': true,
                'slds-form-element_stacked': false,
                'slds-form-element_horizontal': true
            });
        });
    });

    it('should add slds-assistive-text to legend when variant=label-hidden', () => {
        const element = createCheckboxGroup({
            value: ['option1'],
            variant: 'label-hidden',
            required: true,
            disabled: true
        });

        return Promise.resolve().then(() => {
            const hiddenLegend = element.shadowRoot.querySelector(
                'legend.slds-assistive-text'
            );

            expect(hiddenLegend).not.toBeNull();
        });
    });
});
