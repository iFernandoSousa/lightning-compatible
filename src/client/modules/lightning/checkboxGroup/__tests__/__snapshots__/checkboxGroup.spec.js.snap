// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-checkbox-group default 1`] = `
<c-checkbox-group
  class="slds-form-element"
>
  #shadow-root(open)
    <fieldset
      aria-required="false"
    >
      <legend
        class="slds-form-element__legend slds-form-element__label"
      >
        testGroup
      </legend>
      <div
        class="slds-form-element__control"
      >
        <span
          class="slds-checkbox"
        >
          <input
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option1"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Ross
            </span>
          </label>
        </span>
        <span
          class="slds-checkbox"
        >
          <input
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option2"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Rachel
            </span>
          </label>
        </span>
      </div>
    </fieldset>
</c-checkbox-group>
`;

exports[`c-checkbox-group disabled 1`] = `
<c-checkbox-group
  class="slds-form-element"
>
  #shadow-root(open)
    <fieldset
      aria-required="false"
    >
      <legend
        class="slds-form-element__legend slds-form-element__label"
      >
        testGroup
      </legend>
      <div
        class="slds-form-element__control"
      >
        <span
          class="slds-checkbox"
        >
          <input
            disabled=""
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option1"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Ross
            </span>
          </label>
        </span>
        <span
          class="slds-checkbox"
        >
          <input
            disabled=""
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option2"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Rachel
            </span>
          </label>
        </span>
      </div>
    </fieldset>
</c-checkbox-group>
`;

exports[`c-checkbox-group one selected option 1`] = `
<c-checkbox-group
  class="slds-form-element"
>
  #shadow-root(open)
    <fieldset
      aria-required="false"
    >
      <legend
        class="slds-form-element__legend slds-form-element__label"
      >
        testGroup
      </legend>
      <div
        class="slds-form-element__control"
      >
        <span
          class="slds-checkbox"
        >
          <input
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option1"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Ross
            </span>
          </label>
        </span>
        <span
          class="slds-checkbox"
        >
          <input
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option2"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Rachel
            </span>
          </label>
        </span>
      </div>
    </fieldset>
</c-checkbox-group>
`;

exports[`c-checkbox-group required and disabled 1`] = `
<c-checkbox-group
  class="slds-form-element"
>
  #shadow-root(open)
    <fieldset
      aria-required="true"
    >
      <legend
        class="slds-form-element__legend slds-form-element__label"
      >
        <abbr
          class="slds-required"
          title="required"
        >
          *
        </abbr>
        testGroup
      </legend>
      <div
        class="slds-form-element__control"
      >
        <span
          class="slds-checkbox"
        >
          <input
            disabled=""
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option1"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Ross
            </span>
          </label>
        </span>
        <span
          class="slds-checkbox"
        >
          <input
            disabled=""
            id="[shadow:guid]"
            name="testGroup"
            type="checkbox"
            value="option2"
          />
          <label
            class="slds-checkbox__label"
            for="[shadow:guid]"
          >
            <span
              class="slds-checkbox_faux"
            />
            <span
              class="slds-form-element__label"
            >
              Rachel
            </span>
          </label>
        </span>
      </div>
    </fieldset>
</c-checkbox-group>
`;
